design/operator-architecture.asciidoc

```diff
@@ -0,0 +1,89 @@
+# Keycloak Operator High Level Architecture
+
+* **Status**: Draft #1
+* **JIRA**: https://issues.redhat.com/browse/KEYCLOAK-10319[KEYCLOAK-10319]
+* **Source of the images**: https://docs.google.com/presentation/d/13N5ClXXXcxKjXgor72rUdtMBN07_GneUkhTTwBsy_tE/edit?usp=sharing[Google Slides]
+
+This document shows a big picture how all the components and layers of Keycloak Operator work together. The blueprint presented here shall be used a guideline for designing features as well as integrating community Pull Requests.
+
+## Manifesto
+
+We envision Keycloak Operator as:
+
+* an opinionated way of installing and maintaining Keycloak installation on both Kubernetes and OpenShift.
+* an example of a recommended way of installing, configuring, managing custom extension and themes.
+* the thinnest possible layer on top of Keycloak Container and Keycloak Server.
+
+## Introduction
+
+Keycloak Operator was re-designed from the scratch by a small team withing RHMI and RHSSO groups. The implementation plan was described in the [Initial Operator Design Document](operator.md). At the end of a 3 sprint cycle, the team delivered a Level 4 Operator, that was pushed into the [Operatorhub](https://operatorhub.io/operator/keycloak-operator). Once the initial implementation was done, the Operator was handed over to the Keycloak/RHSSO Team.
+
+After maintaining the operator for several months, it became necessary to create a High Level Design (and a long-term vision) how all Keycloak components work together.
+
+## Layered architecture
+
+Keycloak ecosystem consists of 3 layers:
+
+* https://github.com/keycloak/keycloak[Keycloak Server]
+* https://github.com/keycloak/keycloak-containers/tree/master/server[Keycloak Container Image]
+* https://github.com/keycloak/keycloak-operator[Keycloak Operator]
+
+image::img/Keycloak-operator-layers.png[Keycloak ecosystem layers]
+
+Keycloak Operator uses Keycloak Container image, which in turn uses Keycloak Server bits. The goal is to implement features in the lowest possible layer, so that they get inherited by the all upper layers.
+
+.Feature implementation example
+****
+Let's consider a simple feature - adding JSON formatted logs based on an environmental variable.
```

Comment:
TBH When I go through this example, I don't have much clue why the "correct place for the implementation is the container." .

First when looking at the feature description "adding JSON formatted logs based on an environmental variable." - it is not clear to me what exactly this means and which logs are we talking about? Is it possible to add some more context about this?

Then in another sentence, when mentioned this "Implementing it in the Server doesn’t make much sense as it will require manipulating standalone.xml file based on some environment property." I don't understand this TBH...

Just a note that I am noob, who don't have any practical experience with the operators in general. But usually when I want to understand some technology, the examples/quickstarts is the first place, which I am looking at. And from this example I am not able to understand much :) 

Is it possible to add some more context? Or eventually add some links with the expected knowledge for users who look at this design what they need to understand first?

Just my 2 cents when looking at this design from the operator noob perspective :)

Reply:
Ok, I rephrased it. Please give it another look.